### ISSUE TITLE
Overconstrained type variable

### ISSUE THREAD

**Issue by efahl** [python/mypy#5775]
Python 3.6.6 x64
Windows 10 x64
mypy 0.630
mypy-extensions 0.4.1
typing 3.6.6
typing-extensions 3.6.6

https://mail.python.org/pipermail/python-ideas/2018-October/054134.html

Ivan Levkivskyi's suggested solution to handling a "PathLike" protocol:
```python
from typing import AnyStr
from typing_extensions import Protocol
class PathLike(Protocol[AnyStr]):
    def __fspath__(self) -> AnyStr: ...
```
Produces an error unexpected error:
```
> mypy fsp.py
fsp.py:3: error: Invariant type variable 'AnyStr' used in protocol where covariant one is expected
```
Ivan's response
> Hm, it looks like mypy overreacts here. I think it should be safe to use a constrained type variable if there are no constraints that are subtypes of other constraints (which is the case for AnyStr on Python 3, where bytes is not a subtype of str).

**Comment by ilevkivskyi** [comment]
(Edited the repro to be closer to what is actually wanted.)

**Comment by jmehnle** [comment]
Is there a workaround for this? I'm having a similar problem:

```python
from typing import Mapping, Sequence, TypeVar, Union
from typing_extensions import Protocol

Row = TypeVar("Row", bound=Union[Sequence, Mapping])

class C(Protocol[Row]):
    ...
```

`mypy` complains:

```
mypytest.py:6: error: Invariant type variable 'Row' used in protocol where covariant one is expected
```

Any (perhaps dirty) way to make this work?

**Comment by hauntsaninja** [comment]
Make your type variable covariant. (I don't think your case falls under what Ivan's talking about in this issue)

**Comment by Prometheus3375** [comment]
I also have such a trouble with mypy 0.950 and Python 3.9.

```python
from collections.abc import Iterable
from typing import Protocol, TypeVar, overload

K = TypeVar('K')
V = TypeVar('V')


class SupportsKeysAndGetItem(Protocol[K, V]):
    def keys(self, /) -> Iterable[K]: ...
    def __getitem__(self, item: K, /) -> V: ...


@overload
def foo() -> dict: ...
@overload
def foo(**kwargs: V) -> dict[str, V]: ...
@overload
def foo(mapping: SupportsKeysAndGetItem[K, V], /) -> dict[K, V]: ...
@overload
def foo(mapping: SupportsKeysAndGetItem[str, V], /, **kwargs: V) -> dict[str, V]: ...
@overload
def foo(cls, iterable: Iterable[tuple[K, V]], /) -> dict[K, V]: ...
@overload
def foo(iterable: Iterable[tuple[str, V]], /, **kwargs: V, ) -> dict[K, V]: ...


def foo(iterable = (), /, **kwargs):
    return dict(iterable, **kwargs)
```
Produces `sandbox.py:8: error: Invariant type variable "V" used in protocol where covariant one is expected`

**Comment by hauntsaninja** [comment]
Closing as per my previous comment

**Comment by Prometheus3375** [comment]
@hauntsaninja I am pretty sure that mypy overcontrains Protocol variables.

Check this case in [mypy-play.net](https://mypy-play.net/).
```python
from typing import Protocol, TypeVar

K = TypeVar('K')
V = TypeVar('V')


class Protocol1(Protocol[K, V]):
    def foo1(self, /) -> K: ...
    def foo2(self, value: K, /) -> V: ...
    def foo3(self, value: V, /) -> K: ...
```
No errors.

If `foo3` is removed, then this error is produced:
```
main.py:7: error: Invariant type variable "V" used in protocol where covariant one is expected  [misc]
```
If `foo3` and `foo2` are removed, then two errors are produced.
```
main.py:7: error: Invariant type variable "K" used in protocol where covariant one is expected  [misc]
main.py:7: error: Invariant type variable "V" used in protocol where covariant one is expected  [misc]
```
If I make `K` and `V` both covariant and remain all methods, then mypy complains about using a covariant type in method arguments:
```
main.py:9: error: Cannot use a covariant type variable as a parameter  [misc]
main.py:10: error: Cannot use a covariant type variable as a parameter  [misc]
```

All in all, once a user introduces methods with Protocol variables as argument types, then error about covariance is not emitted.

I personally do not see any sense in constraining Protocol variable to be covariant (unless there are methods with them). I find OK for arbitrary protocol `Protocol1[bool]` to not be a subtype of `Protocol1[int]`.

**Comment by Prometheus3375** [comment]
I ma wondering why Protocol type variables must be covariant unless there are methods with them as argument types.
